arch/xtensa/core/xtensa-asm2.c

```diff
@@ -28,6 +28,14 @@ Z_EXC_DECLARE(z_xtensa_user_string_nlen);
 static const struct z_exc_handle exceptions[] = {
 	Z_EXC_HANDLE(z_xtensa_user_string_nlen)
 };
+
+#ifdef CONFIG_THREAD_LOCAL_STORAGE
+/*
+ * Per-thread (TLS) variable indicating whether execution is in user mode.
+ */
+__thread uint32_t is_user_mode;
```

Comment:
Thanks to picolibc where thread local storage must be enabled ==> THREADPTR needs to be valid all the time since we need to make sure `_current` can be used during syscall.

Reply:
Manipulating THREADPTR that way breaks support TLS, which was fine with xcc (that does not support TLS), but it is not with GCC. Regarding user thread manipulates it, the consequence would be an user thread trying to access a memory it hasn't access or executing a privileged instruction. In both cases it should generate an exception, that is pretty similar to an user thread trying to call directly a syscall implementation.

Reply:
`is_user_mode` is a thread variable which is on the thread stack. As for THREADPTR, the arch does not have a way to lock it down, as WUR can be used in any user context. Though, RISC-V has a similar implementation.

If the arch can read PS.RING in user mode, this would not be an issue.
Or we can use a syscall to determine whether you are in usermode, but that has way too much overhead.

Reply:
> If the arch can read PS.RING in user mode, this would not be an issue.

I wonder what might need this information? I can't see in the currently present code how it's used.

Reply:
Then it's the `arch_is_user_context()` that I'm wondering about. Is this just for the syscall entry? Why would the kernel code use syscall instead of calling the handler directly?

Reply:
That's what that function is for. If we are in kernel mode, we call the `z_impl_*` directly. Otherwise, it goes through the syscall path via `arch_syscall_invoke*`. `arch_is_user_context` needs to be as fast as possible to avoid adding too much overhead. Ideally, it should be done in user mode and only with a few instructions.

Reply:
OK, I think I'm getting the point here: @jcmvbkbc-issues the Zephyr model is that "syscalls" are just regular kernel functions with some entry/marshalling code.  In a handful of spots, we need to tell dynamically whether or not we're running in kernel mode (though almost always we can tell statically).  That obviously must be done exclusively with tools that are available to the CPU in ring 3, which basically means register state, THREADPTR and VM mappings.

I think I'm softening on this now, I see the point.  But for sure we have to be 100% correct:

* Set THREADPTR affirmatively to the calling thread's TLS block on syscall entry every time (and set is_user_mode false)
* Set THREADPTR to NULL on interrupt/exception entry so we can't be mislead
* Make sure THREADPTR is NULL in all kernel threads.

Basically every time we enter the kernel we need to reset it.  But having done that I guess this isn't so weird after all.  The entry/exit code could be a tiny bit faster if we did it with a VDSO though, and that would give us a bunch of space to store other things for future proofing.  But as long as the conditions above hold I think I withdraw my objection.

Reply:
THREADPTR cannot be NULL in kernel threads (when thread local storage is enabled) as kernel threads can have thread variables too (which requires THREADPTR to work properly).